tests/lib/peers/PeerManager.test.ts

```diff
@@ -4,87 +4,333 @@ import Polykey from "../../../src/lib/Polykey"
+
+    test('can connect securely to another peer and send data back and forth', async () => {
+      // ==== A to B ==== //
+      const peerConnectionAB = peerA.peerManager.connectToPeer(peerB.peerManager.peerInfo.publicKey)
+      expect(peerConnectionAB).not.toEqual(undefined)
+      expect(await peerConnectionAB.pingPeer(5000)).toEqual(true)
+      // ==== B to A ==== //
+      const peerConnectionBA = peerB.peerManager.connectToPeer(peerA.peerManager.peerInfo.publicKey)
+      expect(peerConnectionBA).not.toEqual(undefined)
+      expect(await peerConnectionBA.pingPeer(5000)).toEqual(true)
+    })
+  })
+
+  describe('NAT Traversal via Peer Relay', () => {
+    let tempDirPeerA: string
+    let peerA: Polykey
+
+    let tempDirPeerB: string
+    let peerB: Polykey
+
+    let tempDirPeerC: string
+    let peerC: Polykey
+
+    beforeAll(async () => {
+      const nodes = await setupTwoConnectedNodes()
+      peerA = nodes.nodeA.pk
+      tempDirPeerA = nodes.nodeA.tempDir
+      peerB = nodes.nodeB.pk
+      tempDirPeerB = nodes.nodeB.tempDir
+
+      // open relay connection from peerB to peerA
+      await peerB.peerManager.turnClient.requestRelayConnection(peerA.peerManager.peerInfo.publicKey)
+      peerA.peerManager.updatePeer(peerB.peerManager.peerInfo)
+      peerB.peerManager.updatePeer(peerA.peerManager.peerInfo)
+      // ======== PEER C ======== //
+      // Define temp directory
+      tempDirPeerC = fs.mkdtempSync(`${os.tmpdir}/pktest${randomString()}`)
+
+      // Create keyManager
+      const keyManagerC = new KeyManager(tempDirPeerC, fs)
+      await keyManagerC.generateKeyPair('John Smith', 'john.smith@email.com', 'some passphrase', 1024, true)
+
+      // Initialize polykey
+      peerC = new Polykey(
+        tempDirPeerC,
+        fs,
+        keyManagerC
+      )
+
+      peerA.peerManager.addPeer(peerC.peerManager.peerInfo)
+      peerB.peerManager.addPeer(peerC.peerManager.peerInfo)
+
+      peerC.peerManager.addPeer(peerA.peerManager.peerInfo)
+      // remove connectedAddress from peerC to simulate a blocked connection
+      peerB.peerManager.peerInfo.connectedAddress = undefined
+      peerC.peerManager.addPeer(peerB.peerManager.peerInfo)
+    })
+
+    afterAll(() => {
+      fs.rmdirSync(tempDirPeerA, { recursive: true })
+      fs.rmdirSync(tempDirPeerB, { recursive: true })
+      fs.rmdirSync(tempDirPeerC, { recursive: true })
+    })
+
+    describe('Peer Relay Sharing', () => {
+      test('can clone a vault through a peer relay connection', async done => {
+        // ==== Pull Vault B to C ==== //
+        const vaultName = `Vault-${randomString()}`
+        const vault = await peerB.vaultManager.createVault(vaultName)
+
+        const clonedVault = await peerC.vaultManager.cloneVault(vault.name, peerB.peerManager.peerInfo.publicKey)
+        expect(vault.name).toEqual(clonedVault.name)
+
+        done()
+      })
+
+      test('can clone many vaults through a peer relay connection', async done => {
```

Comment:
Our relay in Matrix OS can do multiple hops. This is only 1 HOP right? That's fine.

Reply:
yeah this is only one hop. I guess when we implement out DHT, it might be a good opportunity to implement multi hop functionality with something like a multi hop relay. Or do you think pk doesnt really need to be multi-hop?

I kind of think whatever MatrixOS can do with relay, polykey should be able to do as well.